### PR TITLE
Add support for optional aws credentials for s3 for ingestion

### PR DESCRIPTION
**Pull request by maytasm3**
Add support for optional cloud (aws, gcs, etc.) credentials for s3 for ingestion

### Description

Cloud InputSource (such as S3, etc.) should be able to support taking in credentials optionally in order to be able to read from multiple buckets/location. This is in order to use different credentials than the default (i.e. the one provided in the common runtime property) and also to be able to use different credentials for each ingestion.

Motivation: 
This address the proposal https://github.com/apache/druid/issues/9305

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by ccaominh** [comment]
Jobs are failing with a compilation error:
```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/travis/build/apache/druid/extensions-contrib/azure-extensions/src/main/java/org/apache/druid/data/input/azure/AzureInputSource.java:[70,5] constructor CloudObjectInputSource in class org.apache.druid.data.input.impl.CloudObjectInputSource<T> cannot be applied to given types;
  required: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>,org.apache.druid.data.input.impl.CloudConfigProperties
  found: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>
  reason: actual and formal argument lists differ in length
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project druid-azure-extensions: Compilation failure
[ERROR] /home/travis/build/apache/druid/extensions-contrib/azure-extensions/src/main/java/org/apache/druid/data/input/azure/AzureInputSource.java:[70,5] constructor CloudObjectInputSource in class org.apache.druid.data.input.impl.CloudObjectInputSource<T> cannot be applied to given types;
[ERROR]   required: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>,org.apache.druid.data.input.impl.CloudConfigProperties
[ERROR]   found: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>
[ERROR]   reason: actual and formal argument lists differ in length
```

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java [hunk 1]
```diff
@@ -46,13 +51,25 @@
   @JsonCreator
   public S3InputSource(
       @JacksonInject ServerSideEncryptingAmazonS3 s3Client,
+      @JacksonInject AmazonS3ClientBuilder amazonS3ClientBuilder,
+      @JacksonInject S3StorageConfig storageConfig,
       @JsonProperty("uris") @Nullable List<URI> uris,
       @JsonProperty("prefixes") @Nullable List<URI> prefixes,
-      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects
+      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects,
+      @JsonProperty("properties") @Nullable CloudConfigProperties cloudConfigProperties
```
Comment (ccaominh) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: This should probably be added to https://druid.apache.org/docs/latest/ingestion/native-batch.html#s3-input-source. Similar comment for the other modified input sources.

File: core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.druid.metadata.PasswordProvider;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+public class CloudConfigProperties
+{
+  @JsonCreator
+  public CloudConfigProperties(
+      @JsonProperty("accessKeyId") @Nullable PasswordProvider accessKeyId,
```
Comment (zachjsh) [inline core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java]: Do we want to support Azure with this too? If so Azure is a bit different; requires an account id, and secret key. Can this be extended for that easily?

Comment (maytasm3) [reply core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java]: Yes. Maybe we can rename the accessKeyId to something more generic. But all the cloud (s3, G, Azure) requires a "user" and a "password" which this should work for all

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java (hunk 1 above)
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: Yep. Will only modify docs for s3

File: core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java (hunk 2 above)
Comment (maytasm3) [reply core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java]: Making a specific configProperties class for each cloudInputSource without sharing inheritance

File: core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import org.apache.druid.metadata.PasswordProvider;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+public class CloudConfigProperties
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java]: Make sense to me. Making a specific configProperties class for each cloudInputSource without sharing inheritance. (For this PR, just the S3 stuff and Azure and G will comes in a later PR).

**Comment by maytasm3** [comment]
@clintropolis 
Addressed comments. Please review again when you have time.
Main changes:
- Made specific s3ConfigProperties class for S3 without sharing inheritance (hence, removed CloudConfigProperties).
- Added s3ConfigProperties to the constructor for S3InputSource in the withSplit method (so that the peons can connect to s3 using the s3ConfigProperties from main worker)
- Made s3Client lazy initialize inside the S3InputSource
- Update docs

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java [hunk 4]
```diff
@@ -76,7 +114,41 @@ protected S3Entity createEntity(InputSplit<CloudObjectLocation> split)
...
+       ImmutableList.of(split.get()),
+       getS3ConfigProperties()
+    );
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(
+        super.hashCode(),
```
Comment (clintropolis) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: nit: this could be one line

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java [hunk 5]
```diff
@@ -19,50 +19,88 @@
...
+          if (amazonS3ClientBuilder != null && storageConfig != null && s3ConfigProperties != null) {
+            if (s3ConfigProperties.isCredentialsConfigured()) {
+              BasicAWSCredentials creds = new BasicAWSCredentials(
+                  s3ConfigProperties.getAccessKeyId().getPassword(),
+                  s3ConfigProperties.getSecretAccessKey().getPassword());
+              amazonS3ClientBuilder.withCredentials(new AWSStaticCredentialsProvider(creds));
+            }
+            return new ServerSideEncryptingAmazonS3(amazonS3ClientBuilder.build(), storageConfig.getServerSideEncryption());
+          } else {
+            return Preconditions.checkNotNull(s3Client, "s3Client");
```
Comment (clintropolis) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: hmm, should the check that one of `s3Client` or `amazonS3ClientBuilder` and `storageConfig ` are not null be done eagerly instead of in the supplier? (i'm not certain either way, just thinking out loud)

File: docs/ingestion/native-batch.md [hunk 6]
```diff
@@ -846,6 +847,15 @@ S3 Object:
 |bucket|Name of the S3 bucket|None|yes|
 |path|The path where data is located.|None|yes|
 
+Properties Object:
+
+|property|description|default|required?|
+|--------|-----------|-------|---------|
+|accessKeyId|S3 access key for this S3 InputSource|None|yes if secretAccessKey is given|
```
Comment (clintropolis) [inline docs/ingestion/native-batch.md]: I think these can just be shortened to `S3 access key` and `S3 secret key`

Comment (maytasm3) [reply docs/ingestion/native-batch.md]: What do you mean? The description or the property itself?

File: docs/development/extensions-core/s3.md [hunk 7]
```diff
@@ -64,6 +64,7 @@ In addition to this you need to set additional configuration, specific for [deep
 ### S3 authentication methods
 
 To connect to your S3 bucket (whether deep storage bucket or source bucket), Druid use the following credentials providers chain
```
Comment (sthetland) [inline docs/development/extensions-core/s3.md]: Not your change but..
 
```suggestion
Druid uses the following credentials provider chain to connect to your S3 bucket (whether a deep storage bucket or source bucket). 
```

File: docs/development/extensions-core/s3.md [hunk 8]
```diff
@@ -64,6 +64,7 @@ In addition to this you need to set additional configuration, specific for [deep
...
+**Note :** *Below authentication chain can be overridden for connecting to source bucket by specifying the [properties](../../ingestion/native-batch.md#s3-input-source) parameter in the ingestionSpec.*
```
Comment (sthetland) [inline docs/development/extensions-core/s3.md]: I think the doc normally uses a carrot to format note-style paragraphs, like the ones on https://druid.apache.org/docs/latest/tutorials/cluster.html#with-zookeeper-on-master for example. 

I'd use that style, or this could just be a regular second sentence in the paragraph. In either case, I'd rephrase a bit: 

> You can override the default credentials provider chain by specifying an access key using [Properties Object parameters](../../ingestion/native-batch.md#s3-input-source) in the ingestionSpec.

File: docs/ingestion/native-batch.md [hunk 9]
```diff
@@ -838,6 +838,7 @@ Sample specs:
 |uris|JSON array of URIs where S3 objects to be ingested are located.|None|`uris` or `prefixes` or `objects` must be set|
 |prefixes|JSON array of URI prefixes for the locations of S3 objects to be ingested.|None|`uris` or `prefixes` or `objects` must be set|
 |objects|JSON array of S3 Objects to be ingested.|None|`uris` or `prefixes` or `objects` must be set|
+|properties|Properties Object for overriding default S3 configuration. See below for more information.|None|No (Default configurations will be use if not given)
```
Comment (sthetland) [inline docs/ingestion/native-batch.md]:
```suggestion
|properties|Properties Object for overriding the default S3 configuration. See below for more information.|None|No (defaults will be used if not given)
```

File: docs/ingestion/native-batch.md [hunk 10]
```diff
@@ -846,6 +847,15 @@ S3 Object:
...
 |path|The path where data is located.|None|yes|
 
+Properties Object:
+
+|property|description|default|required?|
+|--------|-----------|-------|---------|
+|accessKeyId|S3 access key for this S3 InputSource|None|yes if secretAccessKey is given|
+|secretAccessKey|S3 secret key for this S3 InputSource|None|yes if accessKeyId is given|
+
+**Note :** *If any property is not given in the Properties Object, then the current default server value will be use.*
```
Comment (sthetland) [inline docs/ingestion/native-batch.md]:
```suggestion
Default values are used for any property not overridden in the Properties Object.
```
Caveat to my edit suggestion: it says "any property" but really it's just S3 access key that are supported, no? To avoid the premature generalization, could say something like: 

"If you don't provide an access key, the default S3 credentials provider chain is used." (With a link to the relevant section in s3.md).

File: docs/ingestion/native-batch.md (hunk 6 above)
Comment (maytasm3) [reply docs/ingestion/native-batch.md]: Oh i see. Wanted to emphasize that these configs will only be applied to this input source (not the deep storage, etc.)

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java [hunk 11]
```diff
@@ -76,7 +114,41 @@ protected S3Entity createEntity(InputSplit<CloudObjectLocation> split)
   @Override
   public SplittableInputSource<CloudObjectLocation> withSplit(InputSplit<CloudObjectLocation> split)
   {
-    return new S3InputSource(s3Client, inputDataConfig, null, null, ImmutableList.of(split.get()));
+    return new S3InputSource(
+        s3Client.get(),
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: Not sure what is the correct formatting here.
I do see something like below in the code
`
            CloudFilesObject segmentData = new CloudFilesObject(
                segmentPath,
                outFile,
                objectApi.getRegion(),
                objectApi.getContainer()
            );
`

[the remark this answers is not in the document]
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: I think that's the same formatting.

File: core/src/main/java/org/apache/druid/data/input/impl/S3ConfigProperties.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.druid.metadata.PasswordProvider;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * Contains properties for s3 access configuration.
+ * Properties can be specified by ingestionSpec and override system default.
+ */
+public class S3ConfigProperties
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply core/src/main/java/org/apache/druid/data/input/impl/S3ConfigProperties.java]: Oops forgot that. Done.

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java (hunk 4 above)
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: done

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: oops. Totally didn't see the extra whitespace at the beginning of some lines. Fixed. Thanks

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java (hunk 5 above)
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: I think it will be the same either way. Had it this way to keep all relevant logic for determining which s3Client to use together (inside the Suppliers.memorize(). I'm fine either way though.

File: docs/development/extensions-core/s3.md (hunk 7 above)
Comment (maytasm3) [reply docs/development/extensions-core/s3.md]: Done

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java (hunk 5 above)
Comment (jihoonson) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: I guess it may not matter much here since this is just a sanity check which shouldn't happen, but it would be better to fail early as possible. The check in the constructor will be called in the JSON deserialization while other methods will be called after the task is scheduled on a middleManager.

File: docs/development/extensions-core/s3.md (hunk 8 above)
Comment (maytasm3) [reply docs/development/extensions-core/s3.md]: This is more for like a second sentence in the paragraph. Similar to other note on this page.

Comment (maytasm3) [reply docs/development/extensions-core/s3.md]: You can override the default credentials provider chain for connecting to source bucket by specifying an access key and secret key using Properties Object parameters in the ingestionSpec.

File: docs/ingestion/native-batch.md (hunk 9 above)
Comment (maytasm3) [reply docs/ingestion/native-batch.md]: Done

File: docs/ingestion/native-batch.md (hunk 10 above)
Comment (maytasm3) [reply docs/ingestion/native-batch.md]: That is the idea for the future when we support more than just the access key and secret key. Will changed to "If you don't provide an access key, the default S3 credentials provider chain is used." for now and we can change it back in the future.

Comment (maytasm3) [reply docs/ingestion/native-batch.md]: done

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java (hunk 5 above)
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: Moved all the Preconditions.checkNotNull outside

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java [hunk 13]
```diff
@@ -19,50 +19,88 @@
...
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
 public class S3InputSource extends CloudObjectInputSource<S3Entity>
 {
-  private final ServerSideEncryptingAmazonS3 s3Client;
+  // We lazily initialize s3Client to avoid costly s3 operation when we only need S3InputSource for stored information
+  // (such as for task logs) and not for ingestion. (This cost only applies for new s3Client created with
+  // s3ConfigProperties given).
+  private final Supplier<ServerSideEncryptingAmazonS3> s3Client;
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: Done

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSourceProperties.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.druid.metadata.PasswordProvider;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * Contains properties for s3 input source.
+ * Properties can be specified by ingestionSpec which will override system default.
+ */
+public class S3InputSourceProperties
```
Comment (clintropolis) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSourceProperties.java]: this maybe should be named `S3InputSourceConfig` or something to be more consistently named with other config-ish classes (there is not any types of the form `{name}Properties.java` in druid currently, and I associate the word to `java.util.Properties` which this isn't related to)

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3StorageDruidModule.java [hunk 15]
```diff
@@ -166,9 +166,11 @@ public void configure(Binder binder)
     binder.bind(S3TaskLogs.class).in(LazySingleton.class);
   }
 
+  // This provides ServerSideEncryptingAmazonS3.Builder with default configs from Guice injection initially set.
+  // However, this builder can then be modify and have configuration(s) inside
```
Comment (jihoonson) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3StorageDruidModule.java]: modify -> modified?

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java [hunk 16]
```diff
@@ -42,32 +46,66 @@
...
+  // for stored information (such as for task logs) and not for ingestion.
+  // (This cost only applies for new ServerSideEncryptingAmazonS3 created with s3InputSourceProperties given).
+  private final Supplier<ServerSideEncryptingAmazonS3> s3ClientSupplier;
+  @JsonProperty("properties")
+  private final S3InputSourceProperties s3InputSourceProperties;
   private final S3InputDataConfig inputDataConfig;
 
   @JsonCreator
   public S3InputSource(
       @JacksonInject ServerSideEncryptingAmazonS3 s3Client,
```
Comment (jihoonson) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: Do we still need this `s3Client` parameter to reuse the singleton client? If so, please leave a comment about it.

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSourceProperties.java (hunk 14 above)
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSourceProperties.java]: Done

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3StorageDruidModule.java (hunk 15 above)
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3StorageDruidModule.java]: done

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java (hunk 16 above)
Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: Yea..we still need it to use the singleton client for when the properties is not present in the JSON. This will save the cost of initializing the s3client using the s3ClientSupplier since we can reuse the  injected singleton

Comment (maytasm3) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java]: Yea..we still need it to use the singleton client for when the properties is not present in the JSON. This will save the cost of initializing the s3client using the s3ClientSupplier since we can reuse the  injected singleton

**Comment by maytasm3** [comment]
Manual test looks good.
